server/chat-plugins/helptickets-auto.ts

```diff
@@ -0,0 +1,670 @@
+	}
+	if (punishment.modlogCount) {
+		buf.push(`required modlog: ${punishment.modlogCount}`);
+	}
+	if (punishment.isSingleMessage) {
+		buf.push(`for single messages only`);
+	}
+	return buf.join(', ');
+}
+
+function checkAccess(context: Chat.CommandContext | Chat.PageContext) {
+	if (!WHITELIST.includes(context.user.id)) context.checkCan('bypassall');
+}
+
+export function punishmentsFor(type: string) {
+	return settings.punishments.filter(t => t.ticketType === type);
+}
+
+/** Is punishment1 higher than punishment2 on the list? */
+function supersedes(p1: string, p2: string) {
+	return ORDERED_PUNISHMENTS.indexOf(p1) > ORDERED_PUNISHMENTS.indexOf(p2);
+}
+
+export function determinePunishment(
+	ticketType: string, results: Record<string, number>, modlog: ModlogEntry[], isSingleMessage = false
+) {
+	const punishments = punishmentsFor(ticketType);
+	let action: string | null = null;
+	Utils.sortBy(punishments, p => -ORDERED_PUNISHMENTS.indexOf(p.punishment));
+	for (const punishment of punishments) {
+		if (isSingleMessage && !punishment.isSingleMessage) continue;
+		if (punishment.modlogCount && modlog.length < punishment.modlogCount) continue;
+		if (punishment.severity) {
+			let hit = false;
+			for (const type of punishment.severity.type) {
+				if (results[type] < punishment.severity.certainty) continue;
+				hit = true;
+				break;
+			}
+			if (!hit) continue;
+		}
+		if (!action || supersedes(punishment.punishment, action)) {
+			action = punishment.punishment;
+		}
+	}
+	return action;
+}
+
+export function globalModlog(action: string, user: User | ID | null, note: string, roomid?: string) {
+	user = Users.get(user) || user;
+	void Rooms.Modlog.write(roomid || 'global', {
+		action,
+		ip: user && typeof user === 'object' ? user.latestIp : undefined,
+		userid: toID(user) || undefined,
+		loggedBy: 'artemis' as ID,
+		note,
+	});
+}
+
+export function addModAction(message: string) {
+	Rooms.get('staff')?.add(`|c|&|/log ${message}`).update();
+}
+
+export async function getModlog(params: {user?: ID, ip?: string, actions?: string[]}) {
+	const search: ModlogSearch = {
+		note: [],
+		user: [],
+		ip: [],
+		action: [],
+		actionTaker: [],
+	};
+	if (params.user) search.user = [{search: params.user, isExact: true}];
+	if (params.ip) search.ip = [{search: params.ip}];
+	if (params.actions) search.action = params.actions.map(s => ({search: s}));
+	const res = await Rooms.Modlog.search('global', search);
+	return res?.results || [];
+}
+
+function closeTicket(ticket: TicketState, msg?: string) {
+	if (!ticket.open) return;
+	ticket.open = false;
+	ticket.active = false;
+	ticket.resolved = {
+		time: Date.now(),
+		by: 'the Artemis AI', // we want it to be clear to end users that it was not a human
+		seen: false,
+		staffReason: '',
+		result: msg || "",
+		note: `Want to learn more about the AI? <a href="https://570628/#post-9056769">Visit the information thread</a>.`,
```

Comment:
Do we automatically turn this into a Smogon link?

Reply:
Yes. The note is HTML.

Reply:
Yes, but does our HTML parser convert a link to `https://570628/#post-9056769` to the corresponding Smogon thread?

Reply:
Oh. Oops. Didn't understand that's what you meant.

Reply:
```suggestion
		note: `Want to learn more about the AI? <a href="https://www.smogon.com/forums/threads/pokémon-showdown-forum-rules-resources-read-here-first.3570628/#post-9056769">Visit the information thread</a>.`,
```